packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js

```diff
@@ -88,175 +93,109 @@ function widgetToBlock( widget ) {
 }
 
 export default function useSidebarBlockEditor( sidebar ) {
-	// TODO: Could/should optimize these data structures so that there's less
-	// array traversal. In particular, setBlocks() is a really hot path.
-
-	const [ state, setState ] = useState( () => initState( sidebar ) );
-
-	const ignoreIncoming = useRef( false );
+	const [ blocks, setBlocks ] = useState( () =>
+		widgetsToBlocks( sidebar.getWidgets() )
+	);
 
 	useEffect( () => {
-		const handler = ( event ) => {
-			if ( ignoreIncoming.current ) {
-				return;
-			}
-
-			switch ( event.type ) {
-				case 'widgetAdded': {
-					const { widgetId } = event;
-					const block = blockToWidget(
-						sidebar.getWidget( widgetId )
-					);
-					setState( ( lastState ) => [ ...lastState, block ] );
-					break;
-				}
-
-				case 'widgetRemoved': {
-					const { widgetId } = event;
-					setState( ( lastState ) =>
-						lastState.filter(
-							( { attributes: { __internalWidgetId } } ) =>
-								__internalWidgetId !== widgetId
-						)
-					);
-					break;
-				}
+		return sidebar.subscribe( ( prevWidgets, nextWidgets ) => {
+			setBlocks( ( prevBlocks ) => {
+				const prevWidgetsMap = new Map(
+					prevWidgets.map( ( widget ) => [ widget.id, widget ] )
+				);
+				const prevBlocksMap = new Map(
+					prevBlocks.map( ( block ) => [
+						getWidgetId( block ),
+						block,
+					] )
+				);
 
-				case 'widgetChanged': {
-					const { widgetId } = event;
-					const blockToUpdate = state.find(
-						( { attributes: { __internalWidgetId } } ) =>
-							__internalWidgetId === widgetId
-					);
-					const updatedBlock = widgetToBlock(
-						sidebar.getWidget( widgetId ),
-						blockToUpdate
-					);
-					setState( ( lastState ) =>
-						lastState.map( ( block ) =>
-							block === blockToUpdate ? updatedBlock : block
-						)
-					);
-					break;
-				}
+				const nextBlocks = nextWidgets.map( ( nextWidget ) => {
+					const prevWidget = prevWidgetsMap.get( nextWidget.id );
 
-				case 'widgetsReordered':
-					const { widgetIds } = event;
+					// Bail out updates.
+					if ( prevWidget && prevWidget === nextWidget ) {
+						return prevBlocksMap.get( nextWidget.id );
+					}
 
-					setState( ( lastState ) => {
-						const blocksByWidgetId = keyBy(
-							lastState,
-							'attributes.__internalWidgetId'
-						);
+					return widgetToBlock( nextWidget );
+				} );
 
-						return widgetIds.map(
-							( widgetId ) => blocksByWidgetId[ widgetId ]
-						);
-					} );
-					break;
-			}
-		};
+				// Bail out updates.
+				if ( isShallowEqual( prevBlocks, nextBlocks ) ) {
+					return prevBlocks;
+				}
 
-		return sidebar.subscribe( handler );
+				return nextBlocks;
+			} );
+		} );
 	}, [ sidebar ] );
 
 	const onChangeBlocks = useCallback(
-		( _nextBlocks ) => {
-			ignoreIncoming.current = true;
+		( nextBlocks ) => {
+			setBlocks( ( prevBlocks ) => {
+				if ( isShallowEqual( prevBlocks, nextBlocks ) ) {
```

Comment:
It happened once in my testing, not sure why though 🤔.

Reply:
IIRC, `useBlockSync` works based on the [reference to the blocks array](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/provider/use-block-sync.js#L188); I don't know if it goes through the array of blocks. It theoretically works because the `block-editor` store is supposed to make sure that references to a slice of blocks don't change unless the blocks actually did change. (To prevent re-renders and stuff.)

It might be worth changing [that line](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/provider/use-block-sync.js#L188) to use `isShawllowEqual` as well. I think it may have been avoided to prevent looping over the entire array of blocks every time the subscription callback fires

Reply:
> It might be worth changing [that line](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/provider/use-block-sync.js#L188) to use `isShawllowEqual` as well. I think it may have been avoided to prevent looping over the entire array of blocks every time the subscription callback fires

Yeah I think that would make sense! I believe it also means that I don't have to do the weird `if (nextBlocks === newBlocks) newBlocks = [...nextBlocks]` thing below.

Want to make an issue/PR?